### DISCUSSION TITLE
Share variable between server action and server component (ex. for form validation)

### DISCUSSION THREAD

**Discussion by orichards** [vercel/next.js#49361] [App Router]
Hello!

For server side rendering we have the possibility to return the same HTML content (form for example) with additional validation messages. How is it possible to re-render a server component with error messages provided by the server action function? For sure I could use the client form component and manually invoke form action and check the results but it is not straightforward as using standard form action property. It would be nice if we could share some request scoped variables between server action and server-side components rendering.

What do you think?

**Comment by brhx** [comment]
AFAICT it's not currently possible to get the returned value of a server action when passing it straight to a `<form action>`. We'd need something like the proposed [`useActionState`](https://twitter.com/dan_abramov/status/1654482455267536896) to get the result of the last executed action in an SSR context. 

```ts
"use client";

function MyFormWithErrors({ action }) {
  const result = useActionState(action); // not done yet
  return (
    <form action={action}>
      {result.errors && ...}
      ...
    </form>
  );
}
```

https://twitter.com/dan_abramov/status/1654523681639366657

**Comment by orichards** [comment]
Thank you, Sam! It is still a client component but `useActionState` seems convenient to use. Hope that it will be implemented soon.

**Comment by DFectuoso** [comment]
Is there any updates to this? Seems hard to use server actions in forms in production without being able to give feedback in failed validations to the user.

**Reply by code-guerilla** [reply]
Right now I am using plain HTML/Tailwind for server action form validation.

form ... className='group/form'  

input required... className='group-invalid/form:border-red-500'

span className='hidden group-invalid/form:block'


Just surround them with angle brackets<>.

Sorry It's hard to type on a phone.

It works like a charm.

**Reply by brhx** [reply]
@code-guerilla what do you mean "Tailwind validation"?

**Reply by code-guerilla** [reply]
> @code-guerilla what do you mean "Tailwind validation"?

I edited my comment  :)

**Reply by brhx** [reply]
Thank you for the explanation! I guess I'm still trying to understand what you mean by "Right now I am using **it** in server components", what is **it** in this context? How are you doing server-side form validation?
